[Skia] Simplify and fix accelerated filters

GeorgesStavracas opened the pull request:
#### 7fb06fd42e78309f1d14e5178ce118eda974c625
<pre>
[Skia] Simplify and fix accelerated filters
<a href="https://bugs.webkit.org/show_bug.cgi?id=272991">https://bugs.webkit.org/show_bug.cgi?id=272991</a>

Reviewed by Carlos Garcia Campos.

The ImageBuffer in results must exist for various tests and pages to
work. Skia creates that on demand, like CoreImage, but that doesn&apos;t
seem to work well. Always call &quot;results.imageBuffer()&quot; when applying
Skia filters - FilterImage::imageBuffer() looks like a simple getter
but it actually creates the ImageBuffer if it doesn&apos;t exist!

Remove the SkPictureRecorder and the SkPicture from the game, they&apos;re
not needed since we can paint using the graphics context of the result
image buffer directly.

* Source/WebCore/platform/SourcesSkia.txt:
* Source/WebCore/platform/graphics/filters/FilterImage.cpp:
(WebCore::FilterImage::memoryCost const):
(WebCore::FilterImage::imageBuffer):
* Source/WebCore/platform/graphics/filters/FilterImage.h:
* Source/WebCore/platform/graphics/filters/skia/FEColorMatrixSkiaApplier.cpp:
(WebCore::FEColorMatrixSkiaApplier::apply const):
* Source/WebCore/platform/graphics/filters/skia/FEGaussianBlurSkiaApplier.cpp:
(WebCore::FEGaussianBlurSkiaApplier::apply const):
* Source/WebCore/platform/graphics/filters/skia/SourceGraphicSkiaApplier.cpp:
(WebCore::SourceGraphicSkiaApplier::apply const):
* Source/WebCore/platform/graphics/skia/FilterImageSkia.cpp: Removed.

Canonical link: <a href="https://commits.webkit.org/277914@main">https://commits.webkit.org/277914@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb57bca48a4439de36caeb32656fbb99e28751bf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48838 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28051 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51805 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51525 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44904 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51143 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33999 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25579 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/39926 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49420 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25702 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42113 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21027 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23165 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43293 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6893 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45096 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43783 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53436 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23889 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20158 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47230 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25152 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42322 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46178 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25960 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6996 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24872 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/filters/skia/FEColorMatrixSkiaApplier.cpp:
```diff
@@ -92,13 +94,12 @@ bool FEColorMatrixSkiaApplier::apply(const Filter&, const FilterImageVector& inp
-    auto* canvas = result.beginRecording();
-    canvas->drawImage(nativeImage->platformImage(), 0, 0, { }, &paint);
-    result.finishRecording();
+    static_cast<GraphicsContextSkia*>(&resultImage->context())->drawCustom([=](SkCanvas& canvas) {
```
Use`&` instead `=` for lambda capture, since it's sync call, we don't want to copy any value.

GeorgesStavracas replied:
No lambdas anymore!

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -487,6 +487,14 @@ void GraphicsContextSkia::setupStrokeSource(SkPaint& paint) const
+void GraphicsContextSkia::drawCustom(std::function<void(SkCanvas&)> drawFunc) const
+{
+    if (!makeGLContextCurrentIfNeeded())
+        return;
```
This might not be enough because of the optimization that checks the rendering purpose, because in the case of filters the purpose is Unspecified. I think we should check when the Skia filters are applied. If they always happen during layer rendering, then we don't need the drawCustom, because skia GL context is already current.

GeorgesStavracas replied:
As discussed, I removed this new function from GraphicsContextSkia as it's not needed.

GeorgesStavracas committed:
[Skia] Simplify and fix accelerated filters
https://bugs.webkit.org/show_bug.cgi?id=272991

Reviewed by Carlos Garcia Campos.

The ImageBuffer in results must exist for various tests and pages to
work. Skia creates that on demand, like CoreImage, but that doesn't
seem to work well. Always call "results.imageBuffer()" when applying
Skia filters - FilterImage::imageBuffer() looks like a simple getter
but it actually creates the ImageBuffer if it doesn't exist!

Remove the SkPictureRecorder and the SkPicture from the game, they're
not needed since we can paint using the graphics context of the result
image buffer directly.

* Source/WebCore/platform/SourcesSkia.txt:
* Source/WebCore/platform/graphics/filters/FilterImage.cpp:
(WebCore::FilterImage::memoryCost const):
(WebCore::FilterImage::imageBuffer):
* Source/WebCore/platform/graphics/filters/FilterImage.h:
* Source/WebCore/platform/graphics/filters/skia/FEColorMatrixSkiaApplier.cpp:
(WebCore::FEColorMatrixSkiaApplier::apply const):
* Source/WebCore/platform/graphics/filters/skia/FEGaussianBlurSkiaApplier.cpp:
(WebCore::FEGaussianBlurSkiaApplier::apply const):
* Source/WebCore/platform/graphics/filters/skia/SourceGraphicSkiaApplier.cpp:
(WebCore::SourceGraphicSkiaApplier::apply const):
* Source/WebCore/platform/graphics/skia/FilterImageSkia.cpp: Removed.

Canonical link: https://commits.webkit.org/277914@main
